### PR TITLE
Update ConstantFolding transformation to support Gather with dynamic input

### PR DESCRIPTION
**Pull request by itikhono**
PR sets `can_be_folded` to true for ShapeOf operations, as originally implemented.

before:
image

after:
image

### Tickets:
 - *101315*

### CODE REVIEW

**Comment by jane-intel** [comment]
I approve if it doesn't significantly affect performance of constant folding pass.

**Comment by jane-intel** [comment]
Please see if this PR affected tests failure.
image

File: src/core/src/pass/constant_folding.cpp [hunk 1]
```diff
@@ -23,9 +25,9 @@ using namespace std;
  *
  * \return true if output can be folded otherwise false.
  */
-const auto is_output_foldable = [](const ov::Output<ov::Node>& output) {
+const auto is_output_foldable = [](const ov::Output<ov::Node>& output, const std::string& key = "can_be_folded") {
     const auto& rt_info = output.get_node()->get_rt_info();
-    return !rt_info.count("can_be_folded") || rt_info.at("can_be_folded").as<bool>();
+    return rt_info.count(key) && rt_info.at(key).as<bool>();
```
Comment (praasz) [inline src/core/src/pass/constant_folding.cpp:30]: Is this check ok?
as for default `key` value is not same as before modification there is no negation in first check `!rt_info.count("can_be_folded")` -> `rt_info.count(key)`

**Comment by itikhono** [comment]
the original `true` value for can_be_folded variable was changed in https://github.com/openvinotoolkit/openvino/pull/13766
@nkogteva could you comment on what caused this change?

File: src/core/src/pass/constant_folding.cpp (hunk 1 above)
Comment (itikhono) [reply src/core/src/pass/constant_folding.cpp]: PR is updated, the comment is not relevant

File: src/core/src/pass/constant_folding.cpp [hunk 2]
```diff
@@ -51,6 +53,76 @@ const auto friendly_name_from = [](const ov::Node& node, const size_t output_cou
...
+        int64_t stub = 1;
+        std::vector<int64_t> shape_with_stubs(rank);
+        for (int64_t i = 0; i < rank; ++i) {
+            if (out_pshape[i].is_dynamic()) {
+                dyn_indices.insert(i);
+                shape_with_stubs[i] = stub;
+            } else {
+                shape_with_stubs[i] = out_pshape[i].get_length();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/core/src/pass/constant_folding.cpp]: PR is updated, the comment is not relevant

**Review by nkogteva — APPROVED** [review]

**Review by jane-intel — APPROVED** [review]
Best approach possible

File: src/core/src/pass/constant_folding.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jane-intel) [reply src/core/src/pass/constant_folding.cpp]: Different approach was chosen. Resolving the comment.

File: src/core/src/pass/constant_folding.cpp (hunk 1 above)
Comment (jane-intel) [reply src/core/src/pass/constant_folding.cpp]: Different approach was chosen. Resolving the comment.

**Review by praasz — APPROVED** [review]

**Pull request merged**